java/c/src/main/java/org/apache/arrow/c/ArrayExporter.java

```diff
@@ -96,8 +97,34 @@ void export(ArrowArray array, FieldVector vector, DictionaryProvider dictionaryP
       }
 
       if (buffers != null) {
+        final long bufferSize = buffers.size();
+        /*
+        * For Variadic types, an additional buffer is kept to store
+        * the size of each variadic buffer since that information
+        * cannot be retrieved in the import component.
+        * Here, the dataBufferReqCount is calculated to determine
+        * the additional number of buffers required.
+        * Also note that if the bufferSize is greater than 2, it means
+        * there is one or more data buffers.
+        * Thus, the dataBufferReqCount is set to 1 to get additional buffer
+        * for to store variadic size buffer.
+        * If it is not the case, the dataBuffer is not present.
+        * According to the spec and C Data interface in C++, there must be
+        * at least 3 data buffers present at the import component.
+        * Thus, the dataBufferReqCount is set to 2 to get additional buffer
+        * for empty dataBuffer and the variadic size buffer.
+        */
+        int dataBufferReqCount = 0;
+        if (vector instanceof BaseVariableWidthViewVector) {
```

Comment:
Carving out special cases all over the place means that the implementation is probably wrong. In this case, it looks like getFieldBuffers is too specialized to IPC/the previous assumption that IPC and C Data used the same buffers is now wrong.

I think for C Data, it would be better to have one new method that simply gives the buffer count for ArrayExporter to allocate the buffers pointers, and do any type specific logic during the export.

Reply:
I couldn't agree more. I was hesitant to change the C Data interface as it is a public API. 

But some thoughts...

In `Data.java` we call the `exportVector(...)` which ultimately calls `ArrowExporter.exprt(...)`. And since we recently updated `ArrowRecordBatch` with `variadicBufferCounts` would it be okay/feasible if we just pass it here by introducing a new function

```
public static void exportVector(BufferAllocator allocator, FieldVector vector, DictionaryProvider provider,
      ArrowArray out, ArrowSchema outSchema, List<Integer> variadicBufferCounts) {
....
}
```

Reply:
We don't have that here.

Reply:
Or rather, who would populate that? 

I don't think the C Data Interface should be called directly by consumers. In any case they were going to break at some point because of this anyways so I'd rather fix it properly

Reply:
Yes I mean we can pass that to the export  as a variable.